bkbit/scripts/gfftranslator.py

```diff
@@ -4,133 +4,346 @@
-    with open(file=file, newline='') as csvfile:
-        reader = csv.DictReader(csvfile)
-        translatedannots = []
-        for row in reader:
-            myannotation = kbmodel.GeneAnnotation(id = row['gene_identifier_prefix'].upper().replace('ENE', 'ene') + ':' + row['gene_local_unique_identifier'],
-                                        symbol = row['name'],
-                                        name = row['name'],
-                                        description = row['description'],
-                                        #referenced_in = row['genome_annotation_label'],
-                                        #referenced_in = kbmodel.GenomeAnnotation(id = 'bican:assembly-' + row['genome_annotation_label']),
-                                        referenced_in = genomeAnnot,
-                                        molecular_type = row['gene_biotype'] if row['gene_biotype'] == 'protein_coding' else 'noncoding', #TODO: confirm if noncoding for the rest of the values or parse them as new enum values)
-                                        source_id = row['gene_local_unique_identifier'],
-                                        #in_taxon = [kbmodel.OrganismTaxon(id=taxon_uri[row['taxon_unique_identifier']])],
-                                        in_taxon = [orgTaxon],
-                                        in_taxon_label = taxon_name[row['taxon_unique_identifier']]
-                                        )
-            translatedannots.append(myannotation)
-    return kbmodel.AnnotationCollection(annotations=translatedannots)
+    Resolves duplicates between two GeneAnnotation objects with the same GeneID based on specific conditions.
+    Conditions: Among all duplicates with the same GeneID, keep one record per GeneID using the following priority:
+        1.	“description” is not NULL and “molecular_type” = ‘protein_coding’
+        2.	“description” is not NULL and “molecular_type” = ‘noncoding’
+        3.	“description” is NULL and “molecular_type” = ‘protein_coding’
+        4.	“description” is NULL and “molecular_type” = ‘noncoding
+
+    Parameters:
+    stored_annotation (GeneAnnotation): The stored gene annotation object.
+    curr_annotation (GeneAnnotation): The current gene annotation object.
+
+    Returns:
+    (GeneAnnotation): The gene annotation object with higher priority.
+
+    """
+    if stored_annotation.description is None and curr_annotation.description is not None:
+        return curr_annotation
+    elif stored_annotation.description is not None and curr_annotation.description is None:
+        return stored_annotation
+    elif stored_annotation.molecular_type == 'protein_coding' and curr_annotation.molecular_type == 'noncoding':
+        return stored_annotation
+    elif stored_annotation.molecular_type == 'noncoding' and curr_annotation.molecular_type == 'protein_coding':
+        return curr_annotation
+    else:
+        return stored_annotation
```

Comment:
Yes there is a valid entry to make it to the final `else`. 
Valid entries are when the `stored_annotation` and `curr_annotation` have the same values for **both** description and molecular type.
i.e. Both have “description” is not NULL and “molecular_type” = ‘protein_coding’ 
i.e. Both have “description” is not NULL and “molecular_type” = ‘noncoding’
i.e. Both have “description” is NULL and “molecular_type” = ‘protein_coding’
i.e. Both have “description” is NULL and “molecular_type” = ‘noncoding'

Reply:
Can you implement `__eq__` and `__ne__` for the `GeneAnnotation` class and raise an exception in this case if `stored_annotation != curr_annotation`?

I think I'm only concerned about the cases where "description" is not NULL for both but differs for both. That case feels like an error in the file to me.

It might be worth asking Lydia if the kind of confusion you describe is expected.

Reply:
Let's start by some background why this is needed. 

The NCBI [pipeline ](https://www.ncbi.nlm.nih.gov/genome/annotation_euk/process/#gnomon) seems to employ more than one method to predict genes and else we can end up situation where multiple genes have the same NBCI gene ID. Which is ok for our purpose, it is up to the user of the raw gff file which version of the gene they want to align against.

What we want is symbol and description to go with the NCBI Gene ID.
The problem is when the gene is Gnomon-predicted models, which doesn't fill in the description, while the Ref-Seq models do have description.

Back to Scott's question: Are there cases where "description" is not NILL for both but differs for both?
At this stage, I am not going to rule this out. So, I agree with Scott  - let's make this an error case.
And when we hit that we will have specific examples to work from